### PR TITLE
update soter/openssl to fix compatibility with new openssl version

### PR DESCRIPTION
**Pull request by Lagovas**
due to [changes](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes) in openssl 1.1.0 was changed:
* using openssl structs allocated in heap instead stack because now their structs are opaque in openssl ([1](https://github.com/cossacklabs/themis/compare/master...Lagovas:lagovas/fix-openssl-usage?expand=1#diff-6f18b74912cce552357c27fe1dd7a0bbR25), [2](https://github.com/cossacklabs/themis/compare/master...Lagovas:lagovas/fix-openssl-usage?expand=1#diff-6f18b74912cce552357c27fe1dd7a0bbR31))
* pass soter fields to functions like `some_func(ctx->field)` instead `some_func(&(ctx->field))` because now they are not public and allocated in heap
* create structs using openssl functions *_create and *_new instead struct variable declaration ([1](https://github.com/cossacklabs/themis/compare/master...Lagovas:lagovas/fix-openssl-usage?expand=1#diff-9b33c02104e00ac6e4c049c70ecbe464R44))
* use getters/setters to access to hidden struct fields (EVP_PKEY_type/EVP_PKEY_base_id, RSA_get0_key, RSA_get0_factors, RSA_get0_crt_params, RSA_set0_key, RSA_set0_factors, RSA_set0_crt_params)

want to ask @secumod to check at least src/soter/openssl/soter_rsa_key.c file where work with rsa keys

### CODE REVIEW

File: src/soter/soter_error.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ typedef int soter_status_t;
 #define SOTER_DEBUG_OUT(message)
 #endif
 
+#define UNUSED(x) (void)(x)
```
Comment (vixentael) [inline src/soter/soter_error.h]: :D

File: src/themis/secure_session.c [hunk 2]
```diff
@@ -826,7 +827,7 @@ static themis_status_t secure_session_finish_client(secure_session_t *session_ct
 		return THEMIS_INVALID_PARAMETER;
 	}
 
-	if (memcmp(proto_message->tag, THEMIS_SESSION_PROTO_TAG, SOTER_CONTAINER_TAG_LENGTH))
+	if (memcmp(proto_message->tag, THEMIS_SESSION_PROTO_TAG, SOTER_CONTAINER_TAG_LENGTH) != 0)
```
Comment (vixentael) [inline src/themis/secure_session.c]: we do compare public data here, right?

File: tests/soter/soter_rand_test.c [hunk 3]
```diff
@@ -186,7 +186,9 @@ void run_soter_rand_tests(void)
 {
 	testsuite_enter_suite("soter rand: api");
 	testsuite_run_test(test_api);
-
-//	testsuite_enter_suite("soter rand: NIST STS (make take some time...)");
-//	testsuite_run_test(test_rand_with_nist);
+// always fail under ci
+#ifndef CIRICLE_TEST
```
Comment (vixentael) [inline tests/soter/soter_rand_test.c]: :D

**Comment by vixentael** [comment]
should close #208 after merge

File: src/themis/secure_session.c (hunk 2 above)
Comment (Lagovas) [reply src/themis/secure_session.c]: actually yes, it's trying to check that it's correct secure session packet or no

File: src/soter/openssl/soter_asym_cipher.c [hunk 4]
```diff
@@ -109,11 +110,8 @@ soter_status_t soter_asym_cipher_cleanup(soter_asym_cipher_t* asym_cipher)
 
 	if (asym_cipher->pkey_ctx)
 	{
-		EVP_PKEY* pkey = EVP_PKEY_CTX_get0_pkey(asym_cipher->pkey_ctx);
-		EVP_PKEY_CTX_free(asym_cipher->pkey_ctx);
-		if(pkey){
-		    EVP_PKEY_free(pkey);
-		}
+        EVP_PKEY_CTX_free(asym_cipher->pkey_ctx);
+        asym_cipher->pkey_ctx = NULL;
```
Comment (secumod) [inline src/soter/openssl/soter_asym_cipher.c]: - indentation
- can you confirm that `EVP_PKEY_CTX_free` also frees "contained" `pkey`?

File: src/soter/openssl/soter_hash.c [hunk 5]
```diff
@@ -40,13 +40,19 @@ soter_status_t soter_hash_init(soter_hash_ctx_t *hash_ctx, soter_hash_algo_t alg
...
+		return SOTER_FAIL;
+	}
+
+	if (EVP_DigestInit_ex(hash_ctx->evp_md_ctx, md, NULL))
 	{
 		return SOTER_SUCCESS;
 	}
 	else
 	{
+        soter_hash_cleanup(hash_ctx);
```
Comment (secumod) [inline src/soter/openssl/soter_hash.c]: indent

File: src/soter/openssl/soter_sym.c [hunk 6]
```diff
@@ -99,15 +101,16 @@ soter_sym_ctx_t* soter_sym_ctx_init(const uint32_t alg,
   ctx->alg=alg;
   uint8_t key_[SOTER_SYM_MAX_KEY_LENGTH];
   size_t key_length_=(alg&SOTER_SYM_KEY_LENGTH_MASK)/8;
-  EVP_CIPHER_CTX_init(&(ctx->evp_sym_ctx));
+  //EVP_CIPHER_CTX_init(ctx->evp_sym_ctx);
+  ctx->evp_sym_ctx = EVP_CIPHER_CTX_new();
```
Comment (secumod) [inline src/soter/openssl/soter_sym.c]: probably should be some failure check here

File: src/soter/openssl/soter_sym.c [hunk 7]
```diff
@@ -128,12 +132,12 @@ soter_sym_ctx_t* soter_sym_aead_ctx_init(const uint32_t alg,
   ctx->alg=alg;
   uint8_t key_[SOTER_SYM_MAX_KEY_LENGTH];
   size_t key_length_=(alg&SOTER_SYM_KEY_LENGTH_MASK)/8;
-  EVP_CIPHER_CTX_init(&(ctx->evp_sym_ctx));
+  ctx->evp_sym_ctx = EVP_CIPHER_CTX_new();
```
Comment (secumod) [inline src/soter/openssl/soter_sym.c]: failure check?

File: src/soter/soter_hmac.c [hunk 8]
```diff
@@ -46,23 +46,24 @@ soter_status_t soter_hmac_init(soter_hmac_ctx_t *hmac_ctx, soter_hash_algo_t alg
 		return SOTER_INVALID_PARAMETER;
 	}
 
+    hmac_ctx->hash_ctx = soter_hash_create(algo);
```
Comment (secumod) [inline src/soter/soter_hmac.c]: indentation?

File: src/soter/soter_hmac.c [hunk 9]
```diff
@@ -46,23 +46,24 @@ soter_status_t soter_hmac_init(soter_hmac_ctx_t *hmac_ctx, soter_hash_algo_t alg
...
-		if (SOTER_SUCCESS != res)
-		{
-			return res;
-		}
-
-		res = soter_hash_update(&(hmac_ctx->hash_ctx), key, key_length);
+		res = soter_hash_update(hmac_ctx->hash_ctx, key, key_length);
 		if (SOTER_SUCCESS != res)
 		{
+            soter_hash_destroy(hmac_ctx->hash_ctx);
```
Comment (secumod) [inline src/soter/soter_hmac.c]: indentation

File: src/soter/openssl/soter_asym_cipher.c (hunk 4 above)
Comment (Lagovas) [reply src/soter/openssl/soter_asym_cipher.c]: https://github.com/openssl/openssl/blob/master/crypto/evp/pmeth_lib.c#L339
and also checked with valgrind (he warned that there was double free when I leave `EVP_PKEY_free(pkey)`)

**Pull request merged**